SCClassLibrary/Common/Control/Server.sc

```diff
@@ -892,7 +945,11 @@ Server {
 	}
 
 	freeAll {
```

Comment:
this seems a bit odd, so needs a comment, or separate methods that make explicit what is happening here.

Something like this would be clearer:
```supercollider
Server {
	
	var <>alwaysFreeRoot = false;

	shouldFreeAll { 
		^this.alwaysFreeRoot or: { clientID == 0 }
	}
	
	freeAll {
		if(this.shouldFreeAll) {
			this.freeSynthsInRootGroup
		} {
			this.freeSynthsInDefaultGroup
		};
		this.sendMsg("/clearSched");
 		this.initTree;
	}
	
	freeSynthsInRootGroup {
		this.sendMsg("/g_freeAll", 0)
	}
	
	freeSynthsInDefaultGroup {
		this.sendMsg("/g_freeAll", this.defaultGroupID)
	}
	
}
```

Reply:
Thanks, good idea! 
I think this would be the best behavior: 
by default, clientID == 0 frees root, unless/until user sets it otherwise
by default, clientID > 0 only frees its defaultGroup, unless/until user sets it otherwise

Then don't hardcode the clientID 0 check, 
only set freeAllFreesRoot to true when booting and clientID == 0.
```
Server {
	
	var <>freeAllFreesRoot = false;

	clientID_ { 
		... 
		if (clientID == 0) { freeAllFreesRoot = true };
	}
	
	freeAll {
		if(this.freeAllFreesRoot) {
			this.freeSynthsInRootGroup
		} {
			this.freeSynthsInDefaultGroup
		};
		this.sendMsg("/clearSched");
 		this.initTree;
	}
	
	freeSynthsInRootGroup {
		this.sendMsg("/g_freeAll", 0)
	}
	
	freeSynthsInDefaultGroup {
		this.sendMsg("/g_freeAll", this.defaultGroupID)
	}

}
```
yes?

Reply:
maybe better:

`var <>freeAllFreesRoot = false;`

and 

`if (clientID != 0) { freeAllFreesRoot = false };`

because the default is better reflected that way.

One problem might be that you have to wait until you are notified until you can switch the freeAllFreesRoot-behavior. And if you have already selected one, it'll override it?

Reply:
just discussed with f0:
If user does not set freeAllFreesRoot, do the default: 
client 0 frees rootnode, all others free their defaultgroup.
if user has set freeAllFreesRoot, always do that, no matter what the clientID is.
yes?
```
var <>freeAllFreesRoot; // default is nil, only use freeAllFreesRoot when it was set

freeAll { 

   var freeRoot = if (freeAllFreesRoot.notNil) { freeAllFreesRoot } { clientID == 0) };
   if (freeRoot, { this.freeSynthsInRootNode } {  this.freeSynthsInDefaultGroup };
   
}
```

Reply:
@adcxyz ok, makes sense. But we need to take care that `freeAllFreesRoot`, if set, isn't changed when the server is booted or rebooted.

Reply:
Quick thoughts...

There are a few issues here:

1. Configurability - I think it's important and achievable that this be quite configurable. We don't want to presume how people will want to use this. (This is probably agreed.)

2. Default behaviour - I agree it is probably sensible in shared situations that Cmd-. (which I maintain is a panic button) and freeAll frees your stuff on the remote server, and that the server owner (we should maybe debate what that means, but probably the client whose machine it runs on if such a client exists) should free everything. Of course this will not stop processes on other machines which are spawning synths on yours, which makes Cmd-. feel different...

3. Best Interface - This worries me the most. I'm not sure instance vars in Server are necessarily the best way to do this, and a good think about how this interacts with CmdPeriod and SystemActions is worth the time IMO. I'm reluctant to make it explicitly about client ID, as that is (at least ideally) a behind the scenes thing. ID 0 will go to the first client to connect, and it may not make sense that that person is the 'owner' as it were. (Consider a shared server with no clients on its machine; which client is ID 0 might be arbitrary.) One approach might be to create a concept of a 'shared' server, as distinct from merely 'remote'. 

Sorry if I'm just muddying the waters with this, but I think it's worth trying to get this optimal.

Reply:
@muellmusik - thanks for commenting, agree on all points.

1. yes, make it fully configurable for all cases
   every client can decide whether cmdPeriod frees RootNode or not (local or global emergency brake)
   -> if it does, CmdPeriod should remake all defaultGroups, so other clients can then continue
 2. default behavior:
   for home client (which asks for clientID right after booting, so nearly always gets 0), 
    default should stay as is, for minimum  user surprise.
    for all other clients, default should be polite - only free my own synths.
   + simple way to make home client equally polite, e.g. 
```s.freeAllFreesRootNode = false;```
   + simple way to enable all clients to be equally rude, e.g.
```s.freeAllFreesRootNode = false;```
   + user choice always overrides any clientID-based defaults.
``` if(freeAllFreesRootNode.notNil) { freeAllFreesRootNode } { clientID == 0 }; ```
3. implementation - what would you find the clearest interface ?

Reply:
> implementation - what would you find the clearest interface ?

Well, one thing that I question is making ```freeAll``` a jack-of-all-trades. It seems to me that freeAll should do what it says on the tin. I'd suggest looking at the the CmdPeriod interaction to get the desired effect, and then maybe a ```freeDefaultGroup``` method or something?

Reply:
> Well, one thing that I question is making freeAll a jack-of-all-trades. It seems to me that freeAll should do what it says on the tin. I'd suggest looking at the the CmdPeriod interaction to get the desired effect, and then maybe a freeDefaultGroup method or something?

OK, good point - you are right, this should be in CmdPeriod. 
I'll make a better proposal real soon :-)

Reply:
OK, lets move this redesign question to a new issue, 
and for the PR: 
go back to the previous defaults:
* freeAll means freeAll (cf Miss May ;-) 
* cmd-period only does freeAll on the local server; nothing on remotes

add infrastructure methods and examples for other behaviours:
* freeMyGroup 
* more methods for making and using defaultGroups for groups of clients
* examples for different client behaviors (e.g. polite/rude remotes) 
   done with CmdPeriod.add({ server.freeMyGroup; etc });

Reply:
@muellmusik @telephon 
- I hope the proposal above is an uncontroversial way forward?
Here are tests for the last commit which implements it, 
so you see it in practice.
```
// tests for added defaultGroups and freeAll
//  behavior with default settings first:
s.options.maxLogins_(1);
s.reboot;
s.plotTree;		// watch to debug
s.defaultGroup; // Group(1)
s.defaultGroups; //  added, as only one client, this is [ Group(1) ]
(dur: inf).play;  // play a test sound
Group(0);   // make another group at root level
s.freeAll;  // sound and added group die, only Group(1) remains

// same with using CmdPeriod:
(dur: inf).play;  // play a test sound
Group(0);   // make another group at root level
CmdPeriod.run; // simulate CmdPeriod key action


// test added support methods with 4 client setup:
s.options.maxLogins_(4);
s.reboot;
s.defaultGroups; // four default groups created for all possible clients:
// ->  [ Group(1), Group(67108865), Group(134217729), Group(201326593) ]

// send defaultGroups for specific clientIDs to scsynth : -> cf plotTree
s.sendDefaultGroupsForIDs([0, 1]); // adds second defaultGroup

// send all default groups 
s.sendDefaultGroups;

// make sure we have nothing in here:
s.tree = nil;
s.freeAll; // <- frees other client's defaultGroups too.


// Because s.freeAll frees other clients groups too,
// we should reconstruct them so others can pick up playing again!
// - easiest option:
s.tree = { s.sendDefaultGroups; };
(dur: inf).play;  // test sound
Group(0);   // make second group at root level

CmdPeriod.run; // simulate CmdPeriod key action
// -> still stops all sounds, then remakes all defaultGroups
s.freeAll; // -> plotTree shows four default groups


// Better in the long run: don't even kill the defaultGroup(s).

// convert s to be more polite 
// first, disable general freeAll:
CmdPeriod.freeServers = false;
s.tree = nil;
s.sendDefaultGroups;
// and add a custom action to CmdPeriod:
~freeMyGroup = { s.freeMyGroup };
CmdPeriod.add(~freeMyGroup);
(dur: inf).play;  // test sound
// simulate second sound from a different client:
(dur: inf, degree: 2, group: s.defaultGroups[1].nodeID).play;

CmdPeriod.run; // simulate CmdPeriod key action
// -> only stops s.defaultGroup, others continue untouched

CmdPeriod.remove(~freeMyGroup); // cleanup


// less polite, but still not that rude:
// stop all sounds, but keep all defaultGroups running

~freeDefaultGroups = { s.freeDefaultGroups };
CmdPeriod.add(~freeDefaultGroups);
(dur: inf).play;  // test sound
// fake sound from  client 1:
(dur: inf, degree: 2, group: s.defaultGroups[1].nodeID).play;

CmdPeriod.run; // simulate CmdPeriod key action
// -> stops sound in all groups, groups remain.

Reply:
I will try to look in the next few days. I had one thought (feel free to ignore): Does it help to do things by specifying our relationship with the server, e.g. something like:

```
Server.owned // I 'own' this server, Cmd-. frees all
Server.shared // I don't own this server, Cmd-. frees only my default group
```

We already have *new and *remote, although remoteless (i.e. can we boot it, do we have shared memory) can now be detected via the NetAddr:isLocal.

Not sure if this makes sense, but thought it worth mentioning.

Reply:
@telephon - thanks, and yes, I think so too :-)

@muellmusik - nice idea, could be a good convenience interface on top the new methods in this PR. Maybe not add this in this PR yet; it could become the next long discussion...

Do you mean these as setters and getters that change behavior?
This could be more general and allow more fine-grained control, 
 by having named modes that create different setups, such as:
```
Server.mode = \owned; // freeAll frees root etc
Server.mode = \shared; // freeAll only frees group
Server.mode = \customShared; // variant of shared with adaptations
// basic mode switching functions would be here, others can be added:
Server.modeFuncs = (
   shared: { ... }, 
   owned: { ... }
);
Server.modeFuncs.put(\newbieMode, { ... })
```

Reply:
Yeah, not sure how fine-grained it needs to be, or switchable after creation (maybe not?), or the best way to access it. I thought about class methods for creation, but that's partly just because I thought *remote was such a great simplification.

Reply:
> @muellmusik:
> Yeah, not sure how fine-grained it needs to be, or switchable after creation (maybe not?), or the best way to access it. 

I just updated the guide file with examples for fine-grained control, i.e. different setup variants that make sense to me. I like the flexibility that one can switch them with existing servers, to try with the ensemble which setup tweaks work best for everyone. 
https://github.com/supercollider/supercollider/pull/3180/commits/491508f78ade894e89ec88ae4849239be369c1f2#diff-549b67f8a60696575567372e2a598b4c

> I thought about class methods for creation, but that's partly just because I thought *remote was such a great simplification.

 - yes, *remote is really nice and succinct :-) maybe some of the setups proposed can get equally good shortcut names, for creation or later config switch.